_test/tests/inc/Utils/PageResolverTest.php

```diff
@@ -0,0 +1,153 @@
+<?php
+
+namespace tests\inc\Utils;
+
+use dokuwiki\Utils\PageResolver;
+
+/**
+ * @todo tests that make use of revisions might be wanted
+ */
+class PageResolverTest extends \DokuWikiTest
+{
+    /**
+     * @return \Generator|array
+     * @see testResolveID
+     */
+    public function provideResolveData()
+    {
+        $data = [
+            // relative current in root
+            ['context', 'page', 'page'],
+            ['context', '.page', 'page'],
+            ['context', '.:page', 'page'],
+
+            // relative current in namespace
+            ['lev1:lev2:context', 'page', 'lev1:lev2:page'],
+            ['lev1:lev2:context', '.page', 'lev1:lev2:page'],
+            ['lev1:lev2:context', '.:page', 'lev1:lev2:page'],
+
+            // relative upper in root
+            ['context', '..page', 'page'],
+            ['context', '..:page', 'page'],
+
+            // relative upper in namespace
+            ['lev1:lev2:context', '..page', 'lev1:page'],
+            ['lev1:lev2:context', '..:page', 'lev1:page'],
+            ['lev1:lev2:context', '..:..:page', 'page'],
+            ['lev1:lev2:context', '..:..:..:page', 'page'],
+
+            // deeper nesting
+            ['lev1:lev2:lev3:context', '..page', 'lev1:lev2:page'],
+            ['lev1:lev2:lev3:context', '..:page', 'lev1:lev2:page'],
+            ['lev1:lev2:lev3:context', '..:..page', 'lev1:page'],
+            ['lev1:lev2:lev3:context', '..:..:page', 'lev1:page'],
+            ['lev1:lev2:lev3:context', '..:..:..page', 'page'],
+            ['lev1:lev2:lev3:context', '..:..:..:page', 'page'],
+            ['lev1:lev2:lev3:context', '..:..:..:..page', 'page'],
+            ['lev1:lev2:lev3:context', '..:..:..:..:page', 'page'],
+
+            // strange and broken ones
+            ['lev1:lev2:context', '....:....:page', 'lev1:lev2:page'],
+            ['lev1:lev2:context', '..:..:lev3:page', 'lev3:page'],
+            ['lev1:lev2:context', '..:..:lev3:..:page', 'page'],
+            ['lev1:lev2:context', '..:..:lev3:..:page:....:...', 'page'],
+
+            // relative to current page
+            ['context', '~page', 'context:page'],
+            ['context', '~:page', 'context:page'],
+            ['lev1:lev2:context', '~page', 'lev1:lev2:context:page'],
+            ['lev1:lev2:context', '~:page', 'lev1:lev2:context:page'],
+
+            // start pages
+            ['context', '.:', 'start'],
+            ['foo:context', '.:', 'foo:start'],
+            ['context', 'foo:', 'foo:start'],
+            ['foo:context', 'foo:', 'foo:start'],
+            ['context', '~foo:', 'context:foo:start'],
+            ['foo:context', '~foo:', 'foo:context:foo:start'],
```

Comment:
What happens with `~:`? Does that just point to a start page?

           ['context', '~:', 'context:start'],
           ['foo:context', '~:', 'foo:context:start'],

Reply:
Good question.  The context itself becomes a start page as soon as it exists. In a real world scenario that will always be the case. So at first a link like that will link to itself. Only when another (higher prio) start page exist it will link to that. I added tests to verify that behaviour.